charge-backend/backend_manager.py

```diff
@@ -395,44 +393,55 @@ async def handle_custom_query_retro_molecule(self, data: dict) -> None:
         if agent is None:
             agent = self.experiment.create_agent_with_experiment_state(
                 task=task,
-                agent_name=f"Query agent for {data['nodeId']}",
                 callback=CallbackHandler(self.websocket),
             )
+            self.retro_synth_context.node_id_to_charge_client[data["nodeId"]] = agent
         else:
             agent.task = task
 
         async def run_and_report():
             result = await agent.run()
             # Report answer
             await self._send_processing_message(result, source="Agent")
+            await self.websocket.send_json({"type": "complete"})
 
         await self.task_manager.run_task(run_and_report())
-        await self.websocket.send_json({"type": "complete"})
 
     async def handle_custom_query_retro_product(self, data: dict) -> None:
         """Handle a query on the reaction (from nodeId to its reactants)."""
         assert self.retro_synth_context is not None
 
         await self._send_processing_message(f"Processing reaction query: {data['query']} for node {data['nodeId']} (as product)")
-        node = self.retro_synth_context.node_ids[data["nodeId"]]
-        if data["nodeId"] not in self.retro_synth_context.node_id_to_charge_client:
-            await self._send_processing_message(f"Molecule `{node.smiles}` has no reaction to query.")
-            await self.websocket.send_json({"type": "complete"})
-            return
 
-        agent = self.retro_synth_context.node_id_to_charge_client[data["nodeId"]]
-        agent.task = Task(
+        task = Task(
             system_prompt="You are a helpful chemical assistant who answers in concise but factual responses. Answer the following query about the reaction.",
             user_prompt=data["query"],
         )
 
+        node = self.retro_synth_context.node_ids[data["nodeId"]]
+        if data["nodeId"] not in self.retro_synth_context.node_id_to_charge_client:
+            # No charge client (likely only AZF was run), add context from tree
+            child_nodes = [nid for nid, p in self.retro_synth_context.parents.items() if p == data["nodeId"]]
+            reactants = [self.retro_synth_context.node_ids[nid] for nid in child_nodes]
+            reactants_str = '\n'.join(reactant.smiles for reactant in reactants)
+            task.system_prompt += f"\n\nThe reaction in question is:\nProduct: {node.smiles}\nReactants:\n{reactants_str}"
+            agent = self.experiment.create_agent_with_experiment_state(
+                task=task,
+                callback=CallbackHandler(self.websocket),
+            )
+            self.retro_synth_context.node_id_to_charge_client[data["nodeId"]] = agent
+        else:
+            agent = self.retro_synth_context.node_id_to_charge_client[data["nodeId"]]
+            agent.task = task
+
         async def run_and_report():
             result = await agent.run()
```

Comment:
@tbennun @bvanessen This will ensure results will get added to the shared experiment world  
```suggestion
            result = await agent.run()
            await self.experiment.add_to_context(agent, task, result)
```